xapian-maintainer-tools/profiling/profiler.c

```diff
@@ -0,0 +1,171 @@
+/** @file profiler.c
+ * @brief C file implementing IO logging
+ */
+/* Compile the profiler with:
+ * gcc -shared -fPIC -o profiler.so profiler.c -ldl
+ *
+ * To use it with a process, use:
+ * LD_PRELOAD=/path/to/profiler.so /path/to/executable
+ *
+ * Running this library will produce an strace.log file,
+ * which can be fed to strace-analyse to produce a log, by running:
+ * ./strace-analyse strace.log
+ */
+
+#define _GNU_SOURCE
+
+#include <dlfcn.h>
+#include <fcntl.h>
+#include <string.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+
+// function for logging calls
+
+static int is_reset = 0;
+
+void logcall(const char *format, ...)
+{
+    static FILE *file_ptr;
+    if (!is_reset) {
+	file_ptr = fopen("./strace.log", "w");
+	is_reset = 1;
+    }
+    va_list args_ptr;
+    va_start(args_ptr, format);
+    vfprintf(file_ptr, format, args_ptr);
+    va_end(args_ptr);
+}
```

Comment:
There are a couple of minor problems here.

* If the logfile can't be opened, `file_ptr` will be NULL and the call to `vfprintf()` will likely segfault.  Probably you'd want to skip logging in that case.

* Some of these functions can potentially change `errno` - if the intercepted function we're logging the result of failed and set an `errno` value, and then the logging code changes that to a different one, the calling code will see the intercepted function fail with the wrong `errno` value.  The simplest fix is to wrap `int saved_errno = errno;` ... `errno = saved_errno;` around the code here function.

Another thing I notice if that it's a bit unhelpful to have a fixed log file name, but hard to provide a nice way to specify it here.  We could pass in a filename via an environment variable, but then we don't have a good way to report a failure to open it.

We could instead allow specifying a file descriptor to log to via an environment variable (and default to 2 for `stderr` output by default, which is what tools like `strace` and `ltrace` default to).  Then a `xapian-io-profile` wrapper shell script could open a high file descriptor onto the log file and pass that in, e.g. something like (untested):

```
#!/bin/bash
set -e
case $1 in
  --log=*)
    log_file=`${1/--log=/}`
    shift
    exec 99>"$log_file"
    export XAPIAN_IO_PROFILE_LOG_FD=99
    ;;
esac
LD_PRELOAD=/path/to/profiler.so exec "$@"
```

Then you could profile a command which uses xapian like so, which is a bit neater than having to set `LD_PRELOAD`:

```
xapian-io-profile --log=my-io.log xapian-compact in.db out.db
```

Then if the log file fails to open, the wrapper script reports the error (and the command shouldn't even get run).

Reply:
On IRC you asked:
> wouldn't the script that you wrote(on my PR) require recompilation of profiler.c each time?

No - `XAPIAN_IO_PROFILE_LOG_FD` is an environment variable (which can be read using `getenv()`), not a compiler preprocessor macro.

Reply:
Yes, understood. I implemented it.